File: Java/515.java
__________________________________________________________________________________________________
sample 1 ms submission
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        helper(root, 0, res);
        return res;
    }
    public void helper(TreeNode root, int depth, List<Integer> res){
        if(root == null) return ;
        if(res.size() == depth) {
            res.add(root.val);
        }else{
            res.set(depth, Math.max(root.val, res.get(depth)));
        }
        helper(root.left, depth+1, res);
                    helper(root.right, depth+1, res);
    }
}
__________________________________________________________________________________________________
sample 38012 kb submission
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> largest = new ArrayList<>();
        if (root == null) return largest;
        else {
            helper(root, largest);
            return largest;
        }
    }
    public void helper(TreeNode root, List lst) {
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        while (queue.size() > 0) {
            int max = Integer.MIN_VALUE;
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode removed = queue.poll();
                //System.out.println(removed.val);
                if (removed.val > max) max = removed.val;
                if (removed.left != null) queue.add(removed.left);
                if (removed.right != null) queue.add(removed.right);
            }
            //System.out.println(max);
            lst.add(max);
        }
    }
}
__________________________________________________________________________________________________

File: Python3/515.py
__________________________________________________________________________________________________
sample 40 ms submission
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def largestValues(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        
        queue = collections.deque([root])
        level_largest = []
        while queue:
            level_len = len(queue)
            res = float("-inf")
            while level_len > 0:
                node = queue.popleft()
                if node.val > res:
                    res = node.val
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                level_len -= 1
            level_largest.append(res)
            
        return level_largest
__________________________________________________________________________________________________
sample 15608 kb submission
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def largestValues(self, root: TreeNode) -> List[int]:
        if not root: return []
        maxes = []
        q = [root]
        while q:
            maxes.append(max(n.val for n in q))
            q = [c for n in q for c in (n.left, n.right) if c]
        return maxes
__________________________________________________________________________________________________
